src/pybind/mgr/diskprediction_local/predictor.py

```diff
@@ -39,227 +38,158 @@ class DiskFailurePredictor(object):
+            str -- manufacturer name
         """
+        if model_name.startswith("W"):
+            return "WDC"
+        elif model_name.startswith("T"):
+            return "Toshiba"
+        elif model_name.startswith("S"):
+            return "Seagate"
+        elif model_name.startswith("Hi"):
+            return "Hitachi"
+        else:
+            return "HGST"
 
-        ordered_attrs = []
-
-        for one_day in disk_days:
-            one_day_attrs = []
-
-            for attr in model_attrlist:
-                if attr in one_day:
-                    one_day_attrs.append(one_day[attr])
-                else:
-                    one_day_attrs.append(0)
-
-            ordered_attrs.append(one_day_attrs)
-
-        return ordered_attrs
 
     def predict(self, disk_days):
-        """
-        Predict using given 6-days disk S.M.A.R.T. attributes.
-
-        Args:
-            disk_days: A list struct comprises 6 dictionaries. These
-                       dictionaries store 'consecutive' days of disk SMART
-                       attributes.
-        Returns:
-            A string indicates prediction result. One of following four strings
-            will be returned according to disk failure status:
-            (1) Good : Disk is health
-            (2) Warning : Disk has some symptoms but may not fail immediately
-            (3) Bad : Disk is in danger and data backup is highly recommended
-            (4) Unknown : Not enough data for prediction.
-
-        Raises:
-            Pickle exceptions
-        """
-
-        all_pred = []
-
-        proc_disk_days = self.__preprocess(disk_days)
-        attr_list, diff_data = DiskFailurePredictor.__get_diff_attrs(proc_disk_days)
-        modellist = self.__get_best_models(attr_list)
-        if modellist is None:
-            return "Unknown"
-
-        for modelpath in modellist:
-            model_attrlist = modellist[modelpath]
-            ordered_data = DiskFailurePredictor.__get_ordered_attrs(
-                diff_data, model_attrlist)
-
-            try:
-                with open(modelpath, 'rb') as f_model:
-                    clf = pickle.load(f_model)
+        # massage data into a format that can be fed to models
+        raw_df = self.__format_raw_data(disk_days)
 
-            except UnicodeDecodeError:
-                # Compatibility for python3
-                with open(modelpath, 'rb') as f_model:
-                    clf = pickle.load(f_model, encoding='latin1')
+        # preprocess
+        preprocessed_data = self.__preprocess(raw_df)
+        if preprocessed_data is None:
+            return DiskFailurePredictor.PREDICTION_CLASSES[-1]
 
-            pred = clf.predict(ordered_data)
+        # get model for current manufacturer
+        manufacturer = self.__get_manufacturer(raw_df['model_name'].iloc[0]).lower()
+        model_path = os.path.join(self.model_dirpath, manufacturer + '_predictor.joblib')
+        model = joblib.load(model_path)
```

Comment:
we need to document this part using doc or source. for instance, what classifier was used? and what training parameters were used. what features are used, how are they preprocessed, etc.

as these algorithms are encoded as joblib/pickles, would be better to make it more open/free.

Reply:
I agree. The dataset used, features used, classifiers explored, hyperparams used, etc can be found in the notebooks in [this](https://github.com/chauhankaranraj/ceph_drive_failure) repo. The docs are a bit shabby atm, but the more important notebooks are [dask_data_cleaner.ipynb](https://github.com/chauhankaranraj/ceph_drive_failure/blob/master/dask_data_cleaner.ipynb) and [multiclass_clf.ipynb](https://github.com/chauhankaranraj/ceph_drive_failure/blob/master/multiclass_clf.ipynb).

How do you recommend I add the documentation? Should I write docs for Ceph and mention my repo on it? Or should I directly add notebooks/scripts from my repo to Ceph? Or should I just mention it in the source code, as a comment?

Reply:
Is the goal for others to enhance the code in your repo and eventually the final model will be updated here?

I am also not sure how competing models can be included where on a ceph command line the users can pick their model of choice (i.e. ProphetStor vs Redhat vs Suse etc.)